src/lib.rs

```diff
@@ -90,41 +100,107 @@ pub struct IronOxide {
+    /// Initialization succeeded, and no requests for private key rotations were present
+    NoRotationNeeded(IronOxide),
+    /// Initialization succeeded, but some keys should be rotated
+    RotationNeeded(IronOxide, PrivateKeyRotationCheckResult),
+}
+
+impl InitAndRotationCheck {
+    /// Caller asked to check rotation on initialize, but doesn't want to handle the result.
+    /// Consider using [initialize](fn.initialize.html) instead.
+    pub fn unwrap(self) -> IronOxide {
+        match self {
+            InitAndRotationCheck::NoRotationNeeded(io)
+            | InitAndRotationCheck::RotationNeeded(io, _) => io,
+        }
+    }
+}
+
+/// number of bytes that can be read from the RNG before it is reseeded. 1 MB
+const BYTES_BEFORE_RESEEDING: u64 = 1 * 1024 * 1024;
+
+/// Provides soft rotation capabilities for user and group keys
+pub struct PrivateKeyRotationCheckResult {
+    pub rotations_needed: EitherOrBoth<UserId, Vec<GroupId>>,
+}
+
+impl PrivateKeyRotationCheckResult {
+    pub fn user_rotation_needed(&self) -> Option<UserId> {
+        match &self.rotations_needed {
+            EitherOrBoth::Left(u) | EitherOrBoth::Both(u, _) => Some(u.clone()),
+            _ => None,
+        }
+    }
+
+    //    pub fn group_rotation_needed() -> Option<Vec<GroupId>> {
+    //        unimplemented!()
+    //    }
+
+    //    pub fn rotate_all(ironoxide: &IronOxide) -> (UserId, Vec<GroupId>) //TODO consider when group private key rotation is added
+}
+
 /// Initialize the IronOxide SDK with a device. Verifies that the provided user/segment exists and the provided device
 /// keys are valid and exist for the provided account. If successful returns an instance of the IronOxide SDK
 pub fn initialize(device_context: &DeviceContext) -> Result<IronOxide> {
-    // 1 MB
-    const BYTES_BEFORE_RESEEDING: u64 = 1 * 1024 * 1024;
     let mut rt = Runtime::new().unwrap();
-    let account_id = device_context.account_id();
-    rt.block_on(crate::internal::user_api::user_key_list(
+    rt.block_on(crate::internal::user_api::user_get_current(
         &device_context.auth(),
-        &vec![account_id.clone()],
     ))
-    .and_then(|mut users| {
-        users
-            //We're using remove here because we don't actually need this HashMap anymore and remove
-            //returns us ownership so we can avoid a clone.
-            .remove(&device_context.account_id())
-            .map(|current_user_public_key| IronOxide {
-                recrypt: Recrypt::new(),
-                device: device_context.clone(),
-                user_master_pub_key: current_user_public_key,
-                rng: Mutex::new(ReseedingRng::new(
-                    rand_chacha::ChaChaCore::from_entropy(),
-                    BYTES_BEFORE_RESEEDING,
-                    EntropyRng::new(),
-                )),
-            })
-            .ok_or(IronOxideErr::InitializeError)
+    .and_then(|current_user| {
+        Ok(IronOxide::create_with_side_effects(
```

Comment:
Doesn't seem like you should need the Ok here. and_then just needs something that can go to future and I think a bare value implements that.

Reply:
Hrm. A couple of things here. I don't think bare values implement `IntoFuture`, and this isn't actually a `Future` at all here, it's a `Result`. So `map` can be used to get rid of the `Ok`. I did that.